File: Codes in Python3/Binary Tree/Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        curr=root
        res=[]
        while curr:
            if curr.left is None:
                res.append(curr.val)
                curr=curr.right
            else:
                prev=curr.left
                while prev.right and prev.right!=curr:
                    prev=prev.right
                if prev.right is None:
                    prev.right=curr
                    res.append(curr.val)
                    curr=curr.left
                else:
                    prev.right=None
                    curr=curr.right
        return res
        # s=[]
        # res=[]
        # while root or s:
        #     while root:
        #         res.append(root.val)
        #         s.append(root)
        #         root=root.left
        #     root=s.pop()
        #     root=root.right
        # return res

File: Codes in C++/Binary Tree/Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> preorder;
        TreeNode *curr = root;
        while (curr != NULL)
        {
            if (curr->left == NULL)
            {
                preorder.push_back(curr->val);
                curr = curr->right;
            }
            else
            {
                TreeNode *prev = curr->left;
                while (prev->right && prev->right != curr)
                    prev = prev->right;
                if (prev->right == NULL)
                {
                    prev->right = curr;
                    preorder.push_back(curr->val);
                    curr = curr->left;
                }
                else
                {
                    prev->right = NULL;
                    curr = curr->right;
                }
            }
        }
        return preorder;
    }
    // vector<int> preorderTraversal(TreeNode *root)
    // vector<int> res;
    // stack<TreeNode*> s;
    // while(root!=NULL||!s.empty())
    // {
    //     while(root!=NULL)
    //     {
    //         res.push_back(root->val);
    //         s.push(root);
    //         root=root->left;
    //     }
    //     root=s.top();
    //     s.pop();
    //     root=root->right;
    // }
    // return res;
    // }
};